numalogic/connectors/druid/_druid.py

```diff
@@ -128,33 +158,130 @@ def fetch(
+        df = self._fetch(**query_params)
+
+        if df.empty or df.shape[0] == 0:
+            logging.warning("No data found for keys %s", filter_pairs)
+            return pd.DataFrame()
+
+        df["timestamp"] = pd.to_datetime(df["timestamp"]).astype("int64") // 10**6
+
+        if group_by:
+            df = df.groupby(by=group_by).sum().reset_index()
 
-            _end_time = time.perf_counter() - _start_time
-            _LOGGER.debug("params: %s latency: %.6fs", query_params, _end_time)
-            return df
+        if pivot and pivot.columns:
+            df = df.pivot(
+                index=pivot.index,
+                columns=pivot.columns,
+                values=pivot.value,
+            )
+            df.columns = df.columns.map("{0[1]}".format)
+            df.reset_index(inplace=True)
+
+        _end_time = time.perf_counter() - _start_time
+        _LOGGER.debug("Druid params: %s, fetch time: %.4fs", query_params, _end_time)
+        return df
 
     def raw_fetch(self, *args, **kwargs) -> pd.DataFrame:
         raise NotImplementedError
+
+    def chunked_fetch(
+        self,
+        datasource: str,
+        filter_keys: list[str],
+        filter_values: list[str],
+        dimensions: list[str],
+        delay: float = 3.0,
+        granularity: str = "minute",
+        aggregations: Optional[dict] = None,
+        post_aggregations: Optional[dict] = None,
+        group_by: Optional[list[str]] = None,
+        pivot: Optional[Pivot] = None,
+        hours: int = 24,
+        chunked_hours: int = 6,
+    ) -> pd.DataFrame:
+        """
+        Fetch data concurrently, and concatenate the results.
+
+        Args:
+        ------
+            datasource: Data source to query
+            filter_keys: keys
+            filter_values: values
+            dimensions: The dimensions to group by
+            delay: Added delay to the fetch query from current time.
+            granularity: Time bucket to aggregate data by hour, day, minute, etc.
+            aggregations: A map from aggregator name to one of the
+                ``pydruid.utils.aggregators`` e.g., ``doublesum``
+            post_aggregations: postaggregations map
+            group_by: List of columns to group by
+            pivot: Pivot configuration
+            hours: Hours from now to skip training.
+            chunked_hours: Hours to fetch in each chunk
+
+        Returns
+        -------
+            Fetched dataframe
+
+        Raises
+        ------
+            ValueError: If chunked_hours is less than 1
+        """
+        if chunked_hours < 1:
+            raise ValueError("chunked_hours should be integer and >= 1.")
+
+        _start_time = time.perf_counter()
+        filter_pairs = make_filter_pairs(filter_keys, filter_values)
+
+        hours_elapsed = 0
+        chunked_dfs = []
+        qparams = []
+
+        while hours_elapsed < hours:
+            ref_dt = datetime.now(pytz.utc) - timedelta(hours=hours_elapsed)
```

Comment:
The first call might be redundant here. When hours_elapsed=0, we will be querying druid with the following params: ref_dt = datetime.now(), and the hours = 0 since min(6, 0). So, the first call might be redundant. WDYT? 

Either options:
1. Should not add to qparams if hours=0 
2. Even if we make a query to druid, it is not a big deal since hours=0 anyways.
3. Since, we dont want to add load on druid with real time queries (<2 hours from current time), we might also skip a chunk when that particular chunk requests for latest data (if the timeoverlaps with some point in (currentTime-2 hrs))

Reply:
+1 on the first point. Help us reduce cpu utilization on our end.

I think we still add a delay to every query @shashank10456. So, we are not impacting the real time druid. Every sub-query has a delay param in it if you check. [reference to your point 3]

Reply:
@s0nicboOm should we add delay to every sub-query? For example, if i want a 10 day query, then i should get the data in the range [now - 10days,  now - 2hrs]. If we add chunking to each query, we might miss the pattern or important information. WDYT?

Reply:
Discussed offline, and no change needed here.

Reply:
So, we just pass the endts and the number of hrs of data with this endts.

I think today query works like this:
Let's say 24 hrs of data we are fetching with 6 chunk size.
Original query: [endts = now, hours = 24, delay = 2] = timerange([now - 2, now-26] )

Subquery1: [now, elapsedtime = 0, hours = 6, delay =2] = timerange([now-2, now-8])
Subquery2:[now, elapsedtime = 6, hours = 6, delay =2] = timerange([[now-8, now-14])
Subquery3:[now, elapsedtime = 12, hours = 6, delay =2] = timerange([[now-14, now-20])
Subquery3:[now, elapsedtime = 18, hours = 6, delay =2] = timerange([[now-20, now-26])

@ab93 can correct me if I am wrong.